File: frontend/src/types/voice.ts
// Voice-session wire types + the controller contract VoiceComposer renders.
//
// Two layers here:
//   1. The downstream control frames the ai_service voice WS emits (relayed
//      verbatim by the gateway as `voice:frame`). These mirror
//      backend/ai_service/app/schemas/voice.py 1:1 so the union stays exhaustive.
//   2. `VoiceController` — the small surface VoiceComposer needs, so the component
//      is agnostic to whether it's driven by the Web Speech hook (useVoiceInput,
//      group chat) or the full session loop (useVoiceSession, agent chat).

import type { ExtractedSignals } from './analyze'

// ---- downstream control frames (ai_service -> gateway -> browser) -----------
// Delivered as the socket event `voice:frame` with one of these JSON payloads.

export interface VoiceReadyFrame {
  type: 'ready'
}
export interface VoiceCaptionFrame {
  type: 'caption'
  text: string
  is_final: boolean
}
export interface VoiceBargeInFrame {
  type: 'barge_in'
}
export interface VoiceTurnFinalFrame {
  type: 'turn_final'
  transcript: string
}
export interface VoiceTurnResultFrame {
  type: 'turn_result'
  extracted_signals: ExtractedSignals
  missing_signals: string[]
  agent_reply: string
  is_complete: boolean
  degraded: boolean
}
export interface VoiceSpeechEndFrame {
  type: 'speech_end'
}
export interface VoiceCompleteFrame {
  type: 'complete'
}
export interface VoiceErrorFrame {
  type: 'error'
  message: string
}

// Discriminated union over `type` — narrow on `frame.type` at the receive site.
export type VoiceFrame =
  | VoiceReadyFrame
  | VoiceCaptionFrame
  | VoiceBargeInFrame
  | VoiceTurnFinalFrame
  | VoiceTurnResultFrame
  | VoiceSpeechEndFrame
  | VoiceCompleteFrame
  | VoiceErrorFrame

// ---- upstream control frames (browser -> gateway -> ai_service) -------------
// Sent on the socket event `voice:control`.

export interface VoiceStartFrame {
  type: 'start'
  conversation_history: { role: 'user' | 'assistant'; content: string }[]
  current_signals: ExtractedSignals
}
export interface VoiceMuteFrame {
  type: 'mute'
  muted: boolean
}
export interface VoiceStopFrame {
  type: 'stop'
}
export type VoiceControlFrame = VoiceStartFrame | VoiceMuteFrame | VoiceStopFrame

// ---- the controller VoiceComposer consumes ---------------------------------
// A superset of what useVoiceInput already returns (supported/listening/
// transcript/start/stop/resetTranscript), plus the hands-free-loop fields
// useVoiceSession adds. The extra fields are OPTIONAL so the Web Speech hook (which
// lacks them) still satisfies the type at the group-chat call site.
export interface VoiceController {
  /** Whether voice input is available at all (drives the mic-button render). */
  supported: boolean
  /** The mic is capturing (Web Speech) or the hands-free loop is running. */
  listening: boolean
  /** Live transcript for the in-composer caption. */
  transcript: string
  /** Begin capture / start the hands-free loop. */
  start: () => void
  /** End capture / stop the loop. */
  stop: () => void
  /** Clear the current transcript (used between turns). */
  resetTranscript: () => void

  // --- hands-free-loop extras (useVoiceSession only) ---
  /** The agent is currently speaking (TTS playing) — lets the UI show a state. */
  speaking?: boolean
  /** Mic is muted (still streaming silence upstream to keep Flux alive). */
  muted?: boolean
  /** Toggle mute; no-op for the Web Speech controller. */
  toggleMute?: () => void
  /** 0..1 input level from the worklet RMS, for a real waveform. */
  amplitude?: number
  /** A non-fatal status/error to surface (e.g. "voice busy — showing text only"). */
  error?: string | null
}

File: backend/ai_service/app/schemas/voice.py
"""Control-frame DTOs for the voice WebSocket relay.

The voice WS multiplexes two kinds of frame on one connection:

  * BINARY frames are audio. Upstream (browser -> gateway -> ai_service) they are
    16 kHz mono ``linear16`` mic chunks; downstream they are Cartesia's 44.1 kHz
    ``pcm_f32le`` TTS. Because audio is always the binary frame, no per-frame
    framing header is needed — the transport's own text/binary bit disambiguates.
  * TEXT frames are JSON control messages, modeled here.

Downstream (ai_service -> browser) control frames:
  ready        — handshake done, mic may start.
  caption      — a live interim/final transcript for the on-screen caption.
  barge_in     — the user started talking; the client must stop TTS playback.
  turn_final   — a finalized user turn transcript (before analyze runs).
  turn_result  — the analyzed turn: reconciled signals + reply + completion.
  speech_end   — the current TTS utterance finished streaming.
  complete     — the member has answered everything; the loop may stop.
  error        — a non-fatal problem (e.g. TTS degraded to text-only).

Upstream (browser -> ai_service, forwarded verbatim by the gateway) control frames:
  mute         — {muted}: while muted the client streams silence (Flux has a 60s
                 idle-with-no-keepalive timeout, so a muted mic must not go quiet).
  stop         — end the session; close the upstream Flux stream and TTS.

``turn_result`` intentionally mirrors the ``AnalyzeResponse`` fields the frontend
chat store already consumes, plus a DERIVED ``is_complete`` (there is no such
column on AnalyzeResponse) so the client has one truth for the "all set" state.
"""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field

from app.schemas.ai import ConversationTurn, ExtractedSignals

# ---- downstream (server -> client) -----------------------------------------


class ReadyFrame(BaseModel):
    """Sent once the upstream STT stream is open and the mic may start."""

    type: Literal["ready"] = "ready"


class CaptionFrame(BaseModel):
    """A live transcript fragment for the on-screen caption.

    ``is_final`` marks the confirmed end-of-turn text; interim fragments update
    the caption in place. Captions are display-only — never the analyze input.
    """

    type: Literal["caption"] = "caption"
    text: str
    is_final: bool = False


class BargeInFrame(BaseModel):
    """The user began speaking; the client must stop any TTS playback now."""

    type: Literal["barge_in"] = "barge_in"


class TurnFinalFrame(BaseModel):
    """The finalized user-turn transcript, emitted before analyze runs."""

    type: Literal["turn_final"] = "turn_final"
    transcript: str


class TurnResultFrame(BaseModel):
    """The analyzed turn: reconciled signals + agent reply + completion state.

    ``missing_signals`` is the server-recomputed list (the one the hands-free loop
    trusts); ``is_complete`` is derived from it (empty == complete) so the client
    doesn't re-derive completion differently. ``degraded`` is surfaced so the loop
    can decline to auto-advance / mark complete on a degraded (LLM-failed) turn.
    """

    type: Literal["turn_result"] = "turn_result"
    extracted_signals: ExtractedSignals
    missing_signals: list[str] = Field(default_factory=list)
    agent_reply: str
    is_complete: bool = False
    degraded: bool = False
    # DISPLAY-ONLY expansion of the cuisine lists for the "Noted so far" panel,
    # mirroring AnalyzeResponse. extracted_signals stays COMPACT (the stored/echoed
    # form the latency fix relies on) and the spoken line stays short; these carry
    # the expanded members (asian -> chinese, japanese, ...) the panel renders.
    display_preferred_cuisines: list[str] = Field(default_factory=list)
    display_disliked_cuisines: list[str] = Field(default_factory=list)


class SpeechEndFrame(BaseModel):
    """The current TTS utterance finished streaming (all PCM sent)."""

    type: Literal["speech_end"] = "speech_end"


class CompleteFrame(BaseModel):
    """The member answered everything; the loop may stop capturing."""

    type: Literal["complete"] = "complete"


class ErrorFrame(BaseModel):
    """A non-fatal problem worth surfacing (e.g. TTS degraded to text-only)."""

    type: Literal["error"] = "error"
    message: str


# ---- upstream (client -> server) -------------------------------------------


class StartFrame(BaseModel):
    """Seed the connection with the client's current chat context.

    Sent once, before audio, and re-sendable to re-seed. Carries the SAME history
    and signals the text chat already tracks, so a mixed typed+voice session stays
    coherent: typed turns hit HTTP ``/analyze`` and never reach this handler, so
    without seeding a later voice turn would replay a history missing them and
    re-ask (or drop) a typed answer — dietary especially, since Qa has no dietary
    column and ``_merge_prior_qa`` recovers only cuisines/budget/location.
    """

    type: Literal["start"]
    conversation_history: list[ConversationTurn] = Field(default_factory=list)
    current_signals: ExtractedSignals = Field(default_factory=ExtractedSignals)
    # Cartesia voice id the member picked (settings dropdown). Optional — an
    # absent/unknown id resolves to the server default (see voices.resolve_voice_id).
    # The route parses the frame as a raw dict, so this field documents the wire
    # contract rather than gating it.
    voice_id: str | None = None


class MuteFrame(BaseModel):
    """Toggle mute. While muted the client keeps streaming silence frames.

    NOT a pause: Flux closes an idle ``/v2/listen`` socket after ~60s and accepts
    no KeepAlive message, so the mic stream must keep flowing (zeroed) to survive.
    """

    type: Literal["mute"]
    muted: bool


class StopFrame(BaseModel):
    """End the voice session: close the upstream STT stream and any TTS."""

    type: Literal["stop"]
